hyppo/independence/hhg.py

```diff
@@ -224,3 +309,97 @@ def _pearson_stat(distx, disty):  # pragma: no cover
                     S[i, j] = ((n - 2) * (t12 * t21 - t11 * t22) ** 2) / denom
 
     return S
+
+
+def _point_distance(self, x, y, zx, zy, **kwargs):
+    """
+    For fast HHG,
+    Returns a collection of distances between sample points and chosen centre point
+    """
+    distx = cdist(zx, x, metric=self.compute_distance, **kwargs)
+    disty = cdist(zy, y, metric=self.compute_distance, **kwargs)
+
+    return distx, disty
+
+
+def hoeffdingsD(x, y):
+    """
+    For fast HHG, calculates the Hoeffding's dependence statistic
+    """
+    xin = x
+    yin = y
+    # crop data to the smallest array, length have to be equal
+    if len(xin) < len(yin):
+        yin = yin[: len(xin)]
+    if len(xin) > len(yin):
+        xin = xin[: len(yin)]
+    # dropna
+    x = xin[~(np.isnan(xin) | np.isnan(yin))]
+    y = yin[~(np.isnan(xin) | np.isnan(yin))]
+
+    # undersampling if length too long
+    lenx = len(x)
+    if lenx > 99999:
+        factor = math.ceil(lenx / 100000)
+        x = x[::factor]
+        y = y[::factor]
+
+    R = rankdata(x)
+    S = rankdata(y)
+
+    # core processing
+    N = x.shape
+    dico = {(np.nan, np.nan): np.nan}
+    dicoRin = {np.nan: np.nan}
+    dicoSin = {np.nan: np.nan}
+    dicoRless = {np.nan: np.nan}
+    dicoSless = {np.nan: np.nan}
+    Q = np.ones(N[0])
+
+    i = 0
+    for r, s in np.nditer([R, S]):
+        r = float(r)
+        s = float(s)
+        if (r, s) in dico.keys():
+            Q[i] = dico[(r, s)]
+        else:
+            if r in dicoRin.keys():
+                isinR = dicoRin[r]
+                lessR = dicoRless[r]
+            else:
+                isinR = np.isin(R, r)
+                dicoRin[r] = isinR
+                lessR = np.less(R, r)
+                dicoRless[r] = lessR
+
+            if s in dicoSin.keys():
+                isinS = dicoSin[s]
+                lessS = dicoSless[s]
+            else:
+                isinS = np.isin(S, s)
+                dicoSin[s] = isinS
+                lessS = np.less(S, s)
+                dicoSless[s] = lessS
+
+            Q[i] = (
+                Q[i]
+                + np.count_nonzero(lessR & lessS)
+                + 1 / 4 * (np.count_nonzero(isinR & isinS) - 1)
+                + 1 / 2 * (np.count_nonzero(isinR & lessS))
+                + 1 / 2 * (np.count_nonzero(lessR & isinS))
+            )
+            dico[(r, s)] = Q[i]
+        i += 1
+
+    D1 = np.sum(np.multiply((Q - 1), (Q - 2)))
+    D2 = np.sum(
+        np.multiply(np.multiply((R - 1), (R - 2)), np.multiply((S - 1), (S - 2)))
+    )
+    D3 = np.sum(np.multiply(np.multiply((R - 2), (S - 2)), (Q - 1)))
+
+    D = (
+        30
+        * ((N[0] - 2) * (N[0] - 3) * D1 + D2 - 2 * (N[0] - 2) * D3)
+        / (N[0] * (N[0] - 1) * (N[0] - 2) * (N[0] - 3) * (N[0] - 4))
+    )
+    return D
```

Comment:
this function seems inefficient. Can you send the code where this is ported from?

Reply:
Here's the code:
https://github.com/PaulVanDev/HoeffdingD/blob/master/EfficientHoeffdingD.ipynb

Reply:
I don't know if we can use this, there is no license in the repo

Reply:
yes the first things you said works

Reply:
Editted the function to just be the formula implementation.